Assets/MixedRealityToolkit/Utilities/MixedRealityServiceRegistry.cs

```diff
@@ -0,0 +1,222 @@
+
+        /// <summary>
+        /// Static constructor.
+        /// </summary>
+        static MixedRealityServiceRegistry()
+        { }
+
+        /// <summary>
+        /// Adds an <see cref="IMixedRealityService"/> instance to the registry.
+        /// </summary>
+        /// <typeparam name="T">The interface type of the service being added.</typeparam>
+        /// <param name="serviceInstance">Instance of the service to add.</param>
+        /// <param name="registrar">Instance of the registrar manages the service.</param>
+        /// <returns>
+        /// True if the service was successfully added, false otherwise.
+        /// </returns>
+        public static bool AddService<T>(T serviceInstance, IMixedRealityServiceRegistrar registrar) where T : IMixedRealityService
+        {
+            Type interfaceType = typeof(T);
+            T existingService;
+
+            if (TryGetService<T>(out existingService, serviceInstance.Name))
+            {
+                return false;
+            }
+
+            // Ensure we have a place to put our newly registered service.
+            if (!registry.ContainsKey(interfaceType))
+            {
+                registry.Add(interfaceType, new List<KeyValuePair<IMixedRealityService, IMixedRealityServiceRegistrar>>());
+            }
+
+            List<KeyValuePair<IMixedRealityService, IMixedRealityServiceRegistrar>> services = registry[interfaceType];
+            services.Add(new KeyValuePair<IMixedRealityService, IMixedRealityServiceRegistrar>(serviceInstance, registrar));
+            return true;
+        }
+
+        /// <summary>
+        /// Removes an <see cref="IMixedRealityService"/> instance from the registry.
+        /// </summary>
+        /// <typeparam name="T">The interface type of the service being removed.</typeparam>
+        /// <param name="serviceInstance">Instance of the service to remove.</param>
+        /// <param name="registrar">Instance of the registrar manages the service.</param>
+        /// <returns>
+        /// True if the service was successfully removed, false otherwise.
+        /// </returns>
+        public static bool RemoveService<T>(T serviceInstance, IMixedRealityServiceRegistrar registrar) where T : IMixedRealityService
+        {
+            return RemoveServiceInternal(typeof(T), serviceInstance, registrar);
+        }
+
+        /// <summary>
+        /// Removes an <see cref="IMixedRealityService"/> instance from the registry.
+        /// </summary>
+        /// <typeparam name="T">The interface type of the service being removed.</typeparam>
+        /// <param name="serviceInstance">Instance of the service to remove.</param>
+        /// <returns>
+        /// True if the service was successfully removed, false otherwise.
+        /// </returns>
+        public static bool RemoveService<T>(T serviceInstance) where T : IMixedRealityService
```

Comment:
As a caller, why would I need both of these flavors if it looks like the second one does things automatically for me?) Is it just for effiicency (i.e. some callers know the registrar?)

Reply:
Correct, the overload that takes a registrar parameter allows the code to bypass the loop.